feature_fork_choice_finalization.py is floating

[feature_fork_choice_finalization_fail_log.txt](https://github.com/dtr-org/unit-e/files/2982503/feature_fork_choice_finalization_fail_log.txt)
Failed in travis and I've reproduced it on my machine.

```
2019-03-19 09:15:53.038000 TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "test_framework/test_framework.py", line 282, in start_nodes
    node.wait_for_rpc_connection()
  File "test_framework/test_node.py", line 118, in wait_for_rpc_connection
    assert self.process.poll() is None, "united exited with status %i during initialization" % self.process.returncode
AssertionError: united exited with status 1 during initialization

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_framework/test_framework.py", line 151, in main
    self.setup_network()
  File "./feature_fork_choice_finalization.py", line 45, in setup_network
    self.setup_nodes()
  File "test_framework/test_framework.py", line 235, in setup_nodes
    self.start_nodes()
  File "test_framework/test_framework.py", line 285, in start_nodes
    self.stop_nodes()
  File "test_framework/test_framework.py", line 301, in stop_nodes
    node.stop_node(wait=wait)
  File "test_framework/test_node.py", line 179, in stop_node
    self.drain_main_signal_callbacks_pending()
  File "test_framework/test_node.py", line 188, in drain_main_signal_callbacks_pending
    queue_size = self.getmainsignalscallbackspending()
  File "test_framework/test_node.py", line 94, in __getattr__
    assert self.rpc_connected and self.rpc is not None, "Error: no RPC connection"
AssertionError: Error: no RPC connection
2019-03-19 09:15:53.039000 TestFramework (INFO): Stopping nodes
Traceback (most recent call last):
  File "./feature_fork_choice_finalization.py", line 298, in <module>
    ForkChoiceFinalizationTest().main()
  File "test_framework/test_framework.py", line 175, in main
    self.stop_nodes()
  File "test_framework/test_framework.py", line 301, in stop_nodes
    node.stop_node(wait=wait)
  File "test_framework/test_node.py", line 179, in stop_node
    self.drain_main_signal_callbacks_pending()
  File "test_framework/test_node.py", line 188, in drain_main_signal_callbacks_pending
    queue_size = self.getmainsignalscallbackspending()
  File "test_framework/test_node.py", line 94, in __getattr__
    assert self.rpc_connected and self.rpc is not None, "Error: no RPC connection"
AssertionError: Error: no RPC connection
```

--- a/test/functional/feature_fork_choice_finalization.py
+++ b/test/functional/feature_fork_choice_finalization.py
@@ -191,9 +191,11 @@ class ForkChoiceFinalizationTest(UnitETestFramework):
         payto = finalizer.getnewaddress('', 'legacy')
         txid = finalizer.deposit(payto, 10000)
         wait_until(lambda: self.have_tx_in_mempool([fork1, fork2], txid))
+        disconnect_nodes(finalizer, fork1.index)
 
         # leave instant justification
         fork1.generatetoaddress(3 + 5 + 5 + 5 + 5 + 1, fork1.getnewaddress('', 'bech32'))
+        self.wait_for_vote_and_disconnect(finalizer=finalizer, node=fork1)
         assert_equal(fork1.getblockcount(), 25)
         assert_finalizationstate(fork1, {'currentDynasty': 3,
                                          'currentEpoch': 5,
@@ -213,7 +215,7 @@ class ForkChoiceFinalizationTest(UnitETestFramework):
         # J
         # e5 - e6 fork1, fork2, fork3
         fork1.generatetoaddress(1, fork1.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(fork1.getrawmempool()) == 1, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer, node=fork1)
         fork1.generatetoaddress(4, fork1.getnewaddress('', 'bech32'))
         assert_equal(fork1.getblockcount(), 34)
         assert_finalizationstate(fork1, {'currentDynasty': 4,
@@ -230,6 +232,7 @@ class ForkChoiceFinalizationTest(UnitETestFramework):
         sync_blocks([fork1, fork3])
         disconnect_nodes(fork1, fork3.index)
         fork1.generatetoaddress(1, fork1.getnewaddress('', 'bech32'))
+        self.wait_for_vote_and_disconnect(finalizer=finalizer, node=fork1)
         sync_blocks([fork1, fork2])
 
         for fork in [fork1, fork2]:

--- a/test/functional/feature_fork_choice_forked_finalize_epoch.py
+++ b/test/functional/feature_fork_choice_forked_finalize_epoch.py
@@ -112,8 +112,7 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         #    ..] - [ e6 ] - [ e7 ] - [ e8 ] fork
         disconnect_nodes(node, fork.index)
         fork.generatetoaddress(5 + 5 + 1, fork.getnewaddress('', 'bech32'))
-        connect_nodes(fork, finalizer2.index)
-        wait_until(lambda: len(fork.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer2, node=fork)
         fork.generatetoaddress(1, fork.getnewaddress('', 'bech32'))
         disconnect_nodes(fork, finalizer2.index)
         assert_equal(fork.getblockcount(), 41)
@@ -128,10 +127,8 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         #    |
         #    |                J
         #    ..] - [ e6 ] - [ e7 ] - [ e8 ] fork
-        connect_nodes(node, finalizer1.index)
-
         node.generatetoaddress(1, node.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(node.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer1, node=node)
         node.generatetoaddress(4, node.getnewaddress('', 'bech32'))
         assert_equal(node.getblockcount(), 34)
         assert_finalizationstate(node, {'currentDynasty': 4,
@@ -145,15 +142,13 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         #    |                J
         #    ..] - [ e6 ] - [ e7 ] - [ e8 ] fork
         node.generatetoaddress(1, node.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(node.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer1, node=node)
         node.generatetoaddress(4, node.getnewaddress('', 'bech32'))
         assert_equal(node.getblockcount(), 39)
         assert_finalizationstate(node, {'currentDynasty': 5,
                                         'currentEpoch': 7,
                                         'lastJustifiedEpoch': 6,
                                         'lastFinalizedEpoch': 5})
-
-        disconnect_nodes(node, finalizer1.index)
 
         # test that longer justification doesn't trigger re-org before finalization
         connect_nodes(node, fork.index)
@@ -235,7 +230,7 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         # ... [ e4 ] - [ e5 ] - [ e6 ] node, fork
         connect_nodes(node, finalizer1.index)
         node.generatetoaddress(1, node.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(node.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer1, node=node)
         node.generatetoaddress(4, node.getnewaddress('', 'bech32'))
         sync_blocks([node, fork])
         assert_equal(node.getblockcount(), 34)
@@ -264,9 +259,8 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         #                          |
         #                          |                J
         #                         .. ] - [ e7 ] - [ e8 ] - [ e9 ] fork
-        connect_nodes(fork, finalizer2.index)
         fork.generatetoaddress(1, fork.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(fork.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer2, node=fork)
         fork.generatetoaddress(4, fork.getnewaddress('', 'bech32'))
         assert_equal(fork.getblockcount(), 49)
         assert_finalizationstate(fork, {'currentDynasty': 5,
@@ -282,14 +276,13 @@ class ForkChoiceForkedFinalizeEpochTest(UnitETestFramework):
         #                          |                J
         #                         .. ] - [ e7 ] - [ e8 ] - [ e9 ] fork
         node.generatetoaddress(1, node.getnewaddress('', 'bech32'))
-        wait_until(lambda: len(node.getrawmempool()) > 0, timeout=10)
+        self.wait_for_vote_and_disconnect(finalizer=finalizer1, node=node)
         node.generatetoaddress(1, node.getnewaddress('', 'bech32'))
         assert_equal(node.getblockcount(), 36)
         assert_finalizationstate(node, {'currentDynasty': 5,
                                         'currentEpoch': 7,
                                         'lastJustifiedEpoch': 6,
                                         'lastFinalizedEpoch': 5})
-        disconnect_nodes(node, finalizer1.index)
 
         # node shouldn't switch to fork as it's finalization is behind
         connect_nodes(node, fork.index)

